### ISSUE TITLE
PilotDB.BugData.LifeStages

### ISSUE THREAD

**Issue by philipbaileynar** [namc-utah/namc-bug-db#30]
The following life stages are listed in `PilotDB.TypeLifeStage`:

| LifeStage | Name | Description | Visible |
| :--- | :--- | :--- | :--- |
| A | Adult | NULL | true |
| E | Egg | NULL | true |
| L | Larvae | NULL | true |
| P | Pupae | NULL | true |
| X | Exuvia | NULL | true |

But the following 19 values are found in `PilotDB.BugData.LifeStage`:

| LifeStage | Occurrences |
| :--- | :--- |
| L | 1318086 |
| A | 248451 |
| P | 69097 |
|   | 17005 |
| X | 1556 |
| I | 588 |
| N | 297 |
| E | 165 |
| 0 | 24 |
| 7 | 3 |
| 8 | 3 |
| \] | 2 |
| 6 | 2 |
| d | 2 |
| M | 2 |
| \\ | 1 |
| 9 | 1 |
| 1 | 1 |
| NULL | 0 |

```sql
SELECT LifeStage, Count(LifeStage) Occurrences FROM BugData GROUP BY LifeStage ORDER BY Count(LifeStage) DESC;
```

What should I do with the values in the data that don't exist in the lookup?

**Comment by philipbaileynar** [comment]
Also, @David-Fowler we were planning a **unique** index on the new `BugData` table (`sample.organisms`) was going to be:

```sql
CREATE UNIQUE INDEX ux_organism_organisms ON sample.organisms(sample_id, taxonomy_id, life_stage_id, bug_size);
```

For this to work we need to assign every bug observation with a valid life stage. We can perhaps add a "hidden" life stage called "unknown" for historical samples. This option would not be available through the user interface for new data.

**Comment by Feldspar5** [comment]
Is the 4th record in the long list a blank under lifestage?  I have a black background and do not see anything at all.  I have reached out to my senior taxonomist, but most of those values look like nonsense to me.  I should be able to get a solid answer tomorrow morning.

BTW, you have the only working version of our db.  Ours has been out since Monday.  I might need to start sending queries to you…

**Comment by Feldspar5** [comment]
How about ‘unspecified’ rather than ‘unknown’.

**Comment by philipbaileynar** [comment]
The fourth item in the list is a space.

**Comment by David-Fowler** [comment]
Agreed that should be the index but  bug_size WILL have null values for all samples that do not require measurements. I am pretty sure postgresql allows for nulls in a key / index (sql server does but am questioning myself right now as to the postgres limits).

But agree everything should be associated with a life-stage

**Comment by philipbaileynar** [comment]
I and N have quite a few entries. Everyone sure those aren't legacy life stage types that we should retain but disable for new data?

**Comment by Feldspar5** [comment]
I = immature and is relevant and current
N = ?

I will know tomorrow

**Comment by David-Fowler** [comment]
M I believe was mature historically and yes I believe I and N had names as well.... but I need to look them up if the taxonomists come back with nothing.

**Comment by David-Fowler** [comment]
Ah... I see what was happening. There are some historic samples that have mixed the standardized notes into the LifeStage field. Life Stage is limited to the lookup table @philipbaileynar displayed above but the standardized notes i.e. 

(I) - Immature
(M) - Pour Slide Mount
(T) - Mutation
(R) - Retained in Ref.
(D) - Damaged
(G) - Gender
(U) - Unsure

have been mixed in. So maybe an 'unspecified' like @Feldspar5  noted should be globally applied to all of these discrepancies while perhaps checking the notes field to make sure they include the misplaced standardized note.

**Comment by Feldspar5** [comment]
The taxonomists have no idea what those other values are (from N down, except for E).  Errors?

**Comment by philipbaileynar** [comment]
@Feldspar5 I believe that E is egg.

**Comment by Feldspar5** [comment]
Correct, but that is on top list already I think.
